src/main/java/io/supertokens/totp/Totp.java

```diff
@@ -0,0 +1,318 @@
+            long now = System.currentTimeMillis();
+
+            if (now - latestInvalidCodeCreatedTime < rateLimitResetTimeInMs) {
+                // Less than rateLimitResetTimeInMs (default = 15 mins) time has elasped since
+                // the last invalid code:
+                throw new LimitReachedException(rateLimitResetTimeInMs / 1000);
+
+                // If we insert the used code here, then it will further delay the user from
+                // being able to login. So not inserting it here.
+
+                // Note: One edge case here is: User is rate limited, and then the
+                // DeleteExpiredTotpTokens cron removes the latest invalid attempts
+                // (because they have expired), and then user will again be able to
+                // do extra login attempts (totp_max_attempts more times).
+                // But rate limiting will kick in after totp_max_attempts number
+                // disarming the brute force attack.
+                // Furthermore, the cron running during cooldown of a user is somewhat rare.
+                // So this edge case is practically harmless.
+            }
+        }
+
+        // Check if the code is valid for any device:
+        boolean isValid = false;
+        TOTPDevice matchingDevice = null;
+        for (TOTPDevice device : devices) {
+            // Check if the code is valid for this device:
+            if (checkCode(device, code)) {
+                isValid = true;
+                matchingDevice = device;
+                break;
+            }
+        }
+
+        // Check if the code has been previously used by the user and it was valid (and
+        // is still valid). If so, this could be a replay attack. So reject it.
+        if (isValid) {
+            for (TOTPUsedCode usedCode : usedCodes) {
+                // One edge case is that if the user has 2 devices, and they are used back to
+                // back (within 90 seconds) such that the code of the first device was
+                // regenerated by the second device, then it won't allow the second device's
+                // code to be used until it is expired.
+                // But this would be rare so we can ignore it for now.
+                if (usedCode.code.equals(code) && usedCode.isValid
+                        && usedCode.expiryTime > System.currentTimeMillis()) {
+                    isValid = false;
+                    matchingDevice = null;
+                }
+            }
+        }
+
+        // Insert the code into the list of used codes:
+        long now = System.currentTimeMillis();
+        int invalidCodeExpirySec = Config.getConfig(main).getTotpInvalidCodeExpiryTime(); // (Default 30 mins)
+        int expireInSec = isValid ? matchingDevice.period * (2 * matchingDevice.skew + 1) : invalidCodeExpirySec;
+
+        TOTPUsedCode newCode = new TOTPUsedCode(userId, code, isValid, now + 1000 * expireInSec, now);
+        totpStorage.insertUsedCode(newCode);
```

Comment:
Done. Had to introduce new exception named `UsedCodeAlreadyExistsException` in plugin interface because it's a DB level thing and isn't propagated further.